File: C++/713_SubarrayProductLessThanK.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int left = 0, right = 0, output = 0, product = 1;
        int lleft;
        while(right < nums.size()){
            if (nums[right] > k){
                left = right + 1;
                right++;
                product = 1;
                continue;
            }
            product *= nums[right];
            if (product < k){
                output += right - left + 1;
                right++;
            } else{
                lleft = left;
                while(product >= k && lleft <= right){
                    product /= nums[lleft];
                    lleft++;
                }
                output += right - lleft + 1;
                left = lleft;
                right++;
            } 
        }
        return output;
    }
};

File: 713_SubarrayProductLessThanK.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        left = right = 0
        output = 0
        product = 1
        while right < len(nums):
            if nums[right] >= k:
                left = right + 1
                right += 1
                product = 1
                continue
            product *= nums[right]
            if product < k:
                output += right-left+1
                right += 1
            else:
                lleft = left
                while product >= k and lleft <= right:
                    product /= nums[lleft]
                    lleft += 1
                output += right - lleft + 1
                left = lleft
                right += 1
        return output